metaseq/sequence_generator.py

```diff
@@ -303,12 +407,115 @@ def _sample_topp(self, lprobs):
+                        self.omega_bound,
+                        self.init_p[batch_i, beam_i]
+                        * (self.lambda_decay ** (decay_factor)),
+                    )
+            else:
+                t = toks
+                if self.full_stop_list.to(tokens.device).eq(t).sum() > 0:
+                    self.toks_since_reset[batch_i] = 0
+                else:
+                    self.toks_since_reset[batch_i] += 1
+                decay_factor = max(0, self.toks_since_reset[batch_i] - 1)
+                self.sampling_topp_tensor[batch_i] = torch.max(
+                    self.omega_bound,
+                    self.init_p[batch_i] * (self.lambda_decay ** (decay_factor)),
+                )
+
+    def _apply_repetition_penalties(
+        self,
+        lprobs: torch.Tensor,
+        count_tokens: Optional[torch.Tensor],
+        count_tokens_src: Optional[torch.Tensor],
+    ) -> torch.Tensor:
+        """
+        Apply repetition penalties.
+
+        From https://beta.openai.com/docs/api-reference/engines/retrieve:
+
+        mu[j] -> mu[j] - c[j] * alpha_frequency - float(c[j] > 0) * alpha_presence
+        Where:
+
+        mu[j] is the logit of the j-th token
+        c[j] is how often that token was sampled prior to the current position
+        float(c[j] > 0) is 1 if c[j] > 0 and 0 otherwise
+        alpha_frequency is the frequency penalty coefficient
+        alpha_presence is the presence penalty coefficient
+
+        :param lprobs:
+            (bsz*beam_size, vocab_size) tensor
+        :param count_tokens:
+            (bsz*beam_size, vocab_size) tensor
+        :param count_tokens_src:
+            (bsz*beam_size, vocab_size) tensor
+
+        :return lprobs:
+            return new lprobs!
+        """
+        if self.alpha_frequency > 0 or self.alpha_presence > 0:
+            assert count_tokens is not None
+            lprobs = (
+                lprobs
+                - (count_tokens * self.alpha_frequency)
+                - (count_tokens.gt(0) * self.alpha_presence)
+            )
+        if self.alpha_frequency_src > 0 or self.alpha_presence_src > 0:
+            assert count_tokens_src is not None
+            lprobs = (
+                lprobs
+                - (count_tokens_src * self.alpha_frequency_src)
+                - (count_tokens_src.gt(0) * self.alpha_presence_src)
+            )
+
+        return lprobs
+
+    def _update_repetition_counts(
+        self, tokens: torch.Tensor, count_tokens: Optional[torch.Tensor]
+    ) -> Optional[torch.Tensor]:
+        """
+        Update the repetition counts
+
+        :param tokens:
+            [batchsize * beam, 1]
+        :param count_tokens:
+            [batchsize * beam, vocab_size]
+        """
+        if self.alpha_frequency > 0 or self.alpha_presence > 0:
+            assert count_tokens is not None
+            for i, t in enumerate(tokens):
+                count_tokens[i, t] += 1
```

Comment:
how can we do this on gpu

Reply:
there's probably some fancy indexing I can do, is that preferred to just addition? (is integer addition not on the gpu? I suppose I could also just create a gpu tensor and then do the addition all at once?)

Reply:
Maybe this is fine? This actually never leaves gpu right?

Reply:
it should not